### PR TITLE
Get fallback termination msg from docker when using journald log driver

### LINKED ISSUE #52502: Error when using journald log driver and FallbackToLogsOnError
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

When using the legacy docker container runtime and when a container has `terminationMessagePolicy=FallbackToLogsOnError` and when docker is configured with a log driver other than `json-log` (such as `journald`), the kublet still tries to fetch the last few lines of the container logs from the json log file. When it tries, it gets a message like this:

```
Error on reading termination message from logs: failed to open log file "/var/log/pods/f36cd8db-9984-11e7-9d0f-54e1ad486deb/busybox_0.log": open /var/log/pods/f36cd8db-9984-11e7-9d0f-54e1ad486deb/busybox_0.log: no such file or directory
```

**What you expected to happen**:

The kubelet should not try to get the container's log from the json log file (since it's not there) but should instead ask docker for the logs.

**How to reproduce it (as minimally and precisely as possible)**:

1. Configure `dockerd` to use the `journald` log driver. Add `--log-driver=journald` to its command line.
2. Create a pod with a container that has `terminationMessagePolicy` set to `FallbackToLogsOnError`. For example, put the following into `busybox-pod.yaml`:
```
---
apiVersion: v1
kind: Pod
metadata:
  name: busybox-pod
spec:
  containers:
  - image: gcr.io/google_containers/busybox
    name: busybox
    terminationMessagePolicy: FallbackToLogsOnError
    command: ["/bin/sh"]
    args: ["-c", "echo Last line of container log; exit 1"]
  restartPolicy: Never
```
then run
```
kubectl create -f ~/busybox-pod.yaml
```
3. Wait for the pod to exit with an error
4. Look at the termination message:
```
kubectl get pod -o go-template="{{index .status.containerStatuses 0 "state" "terminated" "message"}}' busybox-pod
```
If you have reproduced the problem, you'll see something like the above error message. If it's working, you'll see
```
Last line of container log
```

**Anything else we need to know?**:
This is related to #42195 from a few months ago.

### PR DESCRIPTION
**Pull request by joelsmith**
**What this PR does / why we need it**:

When using the legacy docker container runtime and when a container has `terminationMessagePolicy=FallbackToLogsOnError` and when docker is configured with a log driver other than `json-log` (such as `journald`), the kubelet should not try to get the container's log from the json log file (since it's not there) but should instead ask docker for the logs.

**Which issue this PR fixes** fixes #52502

**Special notes for your reviewer**:

**Release note**:
```release-note
Fixed log fallback termination messages when using docker with journald log driver
```

### CODE REVIEW

**Comment by sjenning** [comment]
/cc @feiskyer @yujuhong @Random-Liu can we get review on this please?

@derekwaynecarr this is a bug fix when using the legacy docker runtime when docker is using a non json-file logging driver, namely journald.  Should it go in 1.8?

**Comment by derekwaynecarr** [comment]
i am wondering if we can abstract this better and not have to pass the legacy service in

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 1]
```diff
@@ -139,6 +143,7 @@ func NewKubeGenericRuntimeManager(
 	runtimeService internalapi.RuntimeService,
 	imageService internalapi.ImageManagerService,
 	internalLifecycle cm.InternalContainerLifecycle,
+	dockerLegacyService dockershim.DockerLegacyService,
```
Comment (derekwaynecarr) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: can we pass in an interface scoped to what this provides?  for example, `LegacyLogProvider` or something?  this way, we don't need to pass the full legacy service through?

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 2]
```diff
@@ -424,8 +448,13 @@ func (m *kubeGenericRuntimeManager) getPodContainerStatuses(uid kubetypes.UID, n
 			fallbackToLogs := annotatedInfo.TerminationMessagePolicy == v1.TerminationMessageFallbackToLogsOnError && cStatus.ExitCode != 0
 			tMessage, checkLogs := getTerminationMessage(status, annotatedInfo.TerminationMessagePath, fallbackToLogs)
 			if checkLogs {
-				path := buildFullContainerLogsPath(uid, labeledInfo.ContainerName, annotatedInfo.RestartCount)
-				tMessage = m.readLastStringFromContainerLogs(path)
+				// if dockerLegacyService is populated, we're supposed to use it to fetch logs
+				if m.dockerLegacyService != nil {
```
Comment (derekwaynecarr) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: in this area, if we do this, i would prefer we just have a callback function that is passed in if a LegacyLogProvider func is provided?

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 3]
```diff
@@ -397,6 +397,30 @@ func (m *kubeGenericRuntimeManager) readLastStringFromContainerLogs(path string)
 	return buf.String()
 }
 
+// readLastStringFromLegacyContainerLogs attempts to read up to the max log length from the end of the log when the
+// docker runtime is configured to use a log driver other than json-log. It reads up to max log lines.
+func (m *kubeGenericRuntimeManager) readLastStringFromLegacyContainerLogs(uid kubetypes.UID, name, namespace, containerId string) string {
```
Comment (derekwaynecarr) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: this could go in the legacy log provider and kept out of this package.

**Comment by smarterclayton** [comment]
Note that I spotted another error in #52567

**Comment by joelsmith** [comment]
@derekwaynecarr Is this what you had in mind? I haven't squashed yet so that it's easy to compare the two versions. I personally prefer the original version, but am mostly indifferent.

File: pkg/kubelet/dockershim/docker_service.go [hunk 4]
```diff
@@ -524,6 +526,42 @@ func (d *dockerLegacyService) GetContainerLogs(pod *v1.Pod, containerID kubecont
 	return d.client.Logs(containerID.ID, opts, sopts)
 }
 
+// LegacyLogProvider gives the ability to use unsupported docker log drivers (e.g. journald)
+type LegacyLogProvider interface {
```
Comment (derekwaynecarr) [inline pkg/kubelet/dockershim/docker_service.go]: define this interface in the kuberuntime package and not dockershim.

File: pkg/kubelet/dockershim/docker_service.go [hunk 5]
```diff
@@ -524,6 +526,42 @@ func (d *dockerLegacyService) GetContainerLogs(pod *v1.Pod, containerID kubecont
...
+	// Get the last few lines of the logs for a specific container.
+	GetContainerLogTail(uid kubetypes.UID, name, namespace string, containerID kubecontainer.ContainerID) (string, error)
+}
+
+// legacyLogProvider implements the LegacyLogProvider interface
+type legacyLogProvider struct {
+	dls DockerLegacyService
+}
+
+func NewLegacyLogProvider(dls DockerLegacyService) LegacyLogProvider {
```
Comment (derekwaynecarr) [inline pkg/kubelet/dockershim/docker_service.go]: you can keep the implementation here.  if it causes a cycle, i could see an argument to move it in pkg/kubelet, but hopefully that is not the case.

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
this is close to what i was looking for, a few requests.

**Comment by joelsmith** [comment]
@derekwaynecarr I have moved the interface to `kuberuntime` and worked around the cycle. Everything works in my testing. Rebased and squashed.

**Review by derekwaynecarr — APPROVED** [review]

**Pull request merged**